packages/@aws-cdk/core/lib/names.ts

```diff
@@ -36,5 +38,26 @@ export class Names {
     return components.length > 0 ? makeUniqueId(components) : '';
   }
 
+  /**
+   * Returns a CloudFormation-compatible unique identifier for a construct based
+   * on its path. If the node is a Stack, the stack name is used for that path component instead.
+   * The identifier includes a human readable portion rendered from the path components and a hash suffix.
+   * @param construct The construct
+   * @param maxUniqueIdLength The maximum length of the unique id
+   * @returns a unique id based on the construct path
+   */
+  public static uniqueIdWithStackName(construct: Construct, maxUniqueIdLength: number = 256): string {
+    const node = Node.of(construct);
+    const components = node.scopes.slice(1).map(component => {
+      if (Stack.isStack(component) && !unresolved(component.stackName)) {
```

Comment:
If `stackName` is unresolved, this function doesn't help anymore, but feels like we should be able to do something here anyway (by using `{ Ref: 'AWS::StackName' }`). We also don't need the entire construct path anymore, just the path below the stack.

How about we do something like:

```ts
const stack = Stack.of(node);
const components = scopesUnder(node, stack); // Only return the ids of nodes underneath the given stack

const stackNameLength = Token.isUnresolved(stack.stackName) ? 128 : stack.StackName.length;

const uniqueId = components.length > 0 ? makeUniqueId(components) : '';
const startIndex = Math.max(0, uniqueId.length - (maxUniqueIdLength) - stackNameLength - 1);

return `${stack.stackName}-${uniqueId.substring(startIndex)}`;
```

Reply:
In fact we should have something similar to this in calculating `CfnOutput` export names

Reply:
Correct me if I'm wrong but the only case where a stack name can be unresolved is in a nested stack. When I tested creating stacks with names that were unresolved, I got the error message when running `cdk synth`:
```
/Volumes/workplace/experiments/node_modules/@aws-cdk/core/lib/stack.ts:376
    throw new Error(`Stack name must match the regular expression: ${VALID_STACK_NAME_REGEX.toString()}, got '${this.stackName}'`);
```

With nested stacks, the StackName of the top level stack will still be used as the first part of the uniqueId and then the nested stack ID will be used so they will be unique and no unresolved tokens will be used.

I'll add extra test cases to make sure I'm not working on failed assumptions.

Reply:
> the only case where a stack name can be unresolved is in a nested stack

I might be being overly cautious here. I guess I was thinking of the case where a user would overload `stackName` to return `{ Ref: 'AWS::StackName' }`. We might say: "that's not allowed", and that might be fair. But if we can allow that case with a little more work, why not allow it?

A slightly more concrete issue that would be solved: by including the entire construct path in the unique name, moving the Stack in the construct hierarchy becomes a breaking change. If we take the Stack as "root" of the unique name, it doesn't matter where the Stack lives in the construct hierarchy, and we can move it around without affecting the generated names. That seems preferable.

Example:

```
        ┌────┐                               ┌────┐        
        │App │                               │App │        
        └────┘                               └────┘        
           │                                    │          
      ┌────┴────┐                          ┌────┴────┐     
      │         │                          │         │     
   ┌────┐    ┌────┐                     ┌────┐    ┌────┐   
   │ A  │    │ B  │                     │ A  │    │ B  │   
   └────┘    └────┘                     └────┘    └────┘   
      │                                              │     
      │                                              │     
  ┌───────┐                                      ┌───────┐ 
  │ Stack │                                      │ Stack │ 
  └───────┘            ━━━━━━━━━━━━▶             └───────┘ 
      │                                              │     
      │                                              │     
   ┌────┐                                         ┌────┐   
   │Pipe│                                         │Pipe│   
   └────┘                                         └────┘   
      │                                              │     
      │                                              │     
┌──────────┐                                   ┌──────────┐
│ KeyAlias │                                   │ KeyAlias │
└──────────┘                                   └──────────┘
```

If the name of `KeyAlias` depends on `A/Stack/Pipeline/KeyAlias` then it will be renamed if we ever move the Stack to `B`.

If its name just depended on `Stack/Pipeline/KeyAlias` then moving the stack wouldn't matter.